src/hip_stream.cpp

```diff
@@ -254,14 +254,101 @@ hipError_t hipStreamGetPriority(hipStream_t stream, int* priority) {
 
 
 //---
+hipError_t setCallbackPacket(hsa_queue_t* queue,
+                             uint64_t& index, uint64_t& nextIndex,
+                             hsa_barrier_and_packet_t** barrier1,
+                             hsa_barrier_and_packet_t** barrier2){
+
+    uint64_t tempIndex = 0;
+    uint32_t mask = queue->size - 1;
+    hsa_barrier_and_packet_t* tempBarrier;
+
+    // Check for empty barrier packets
+    do{
+
+        tempIndex = hsa_queue_load_write_index_scacquire(queue);
+        tempBarrier = &(((hsa_barrier_and_packet_t*)(queue->base_address))[tempIndex & mask]);
+    }while(!(tempBarrier->header & HSA_PACKET_TYPE_INVALID));
+
+    // Reserve two packets for two barriers
+    index = hsa_queue_add_write_index_scacquire(queue, 2);
+
+    if(index > mask)
+    {
+        index = 0;
+        nextIndex = 1;
+    }
+    else if(index == mask)
+        nextIndex = 0;
+    else
+        nextIndex = index + 1;
+
+    tempBarrier = new hsa_barrier_and_packet_t;
+    memset(tempBarrier, 0, sizeof(hsa_barrier_and_packet_t));
+    tempBarrier->header = HSA_PACKET_TYPE_INVALID;
+
+    // Barrier 1
+    *barrier1 = &(((hsa_barrier_and_packet_t*)(queue->base_address))[index & mask]);
+    memcpy(*barrier1,tempBarrier,sizeof(hsa_barrier_and_packet_t));
+
+    // Barrier 2
+    *barrier2 = &(((hsa_barrier_and_packet_t*)(queue->base_address))[nextIndex & mask]);
+    memcpy(*barrier2,tempBarrier,sizeof(hsa_barrier_and_packet_t));
+
+    delete tempBarrier;
+
+    return hipSuccess;
+}
+
 hipError_t hipStreamAddCallback(hipStream_t stream, hipStreamCallback_t callback, void* userData,
                                 unsigned int flags) {
+
     HIP_INIT_API(hipStreamAddCallback, stream, callback, userData, flags);
     hipError_t e = hipSuccess;
 
-    // Create a thread in detached mode to handle callback
+    // 1. Lock the queue
+    hsa_queue_t* lockedQ = static_cast<hsa_queue_t*> (stream->criticalData()._av.acquire_locked_hsa_queue());
+
+    // 2. Allocate a singals
+    hsa_signal_t signal;
+    hsa_status_t status = hsa_signal_create(1, 0, NULL, &signal);
+
+    hsa_signal_t depSignal;
+    status = hsa_signal_create(1, 0, NULL, &depSignal);
```

Comment:
Here and below: no status checks. Please provide them and their translation to HIP's ones with corresponding returns.